### PR TITLE
Gradients - From Terminal Text Effects

### PR DESCRIPTION
**Pull request by tznind**
I have started looking at [Terminal Text Effects](https://github.com/ChrisBuilds/terminaltexteffects/) python library and did some experimental porting with some Chat GPT help.

I've switched to our Color scheme and Point but the rest is largely unchanged.

Also found a bug in LineCanvas intersections coordinates being wrong on the data model (not actually affecting rendering but came up as I tested the gradient mapping logic).

I initially started porting everything but it got a bit out of hand.  This is just the gradient class ported and a showcase.

image

## Fixes

- Fixes #_____

## Proposed Changes/Todos

- [x] No nested classes
- [x] Tests
- [x] xmldoc

## Pull Request checklist:

- [ ] I've named my PR in the form of "Fixes #issue. Terse description."
- [x] My code follows the [style guidelines of Terminal.Gui](https://github.com/gui-cs/Terminal.Gui/blob/develop/.editorconfig) - if you use Visual Studio, hit `CTRL-K-D` to automatically reformat your files before committing.
- [x] My code follows the [Terminal.Gui library design guidelines](https://github.com/gui-cs/Terminal.Gui/blob/develop/CONTRIBUTING.md)
- [x] I ran `dotnet test` before commit
- [x] I have made corresponding changes to the API documentation (using `///` style comments)
- [x] My changes generate no new warnings
- [x] I have checked my code and corrected any poor grammar or misspellings
- [x] I conducted basic QA to assure all features are working

### CODE REVIEW

**Comment by tig** [comment]
This could be leveraged for

- #2800

**Comment by tznind** [comment]
> This could be leveraged for #2800

Yup we could definetly do some kind of color wheel/rectangle thing


Any idea why CI tests are aborting with?

```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. System.ArgumentException: There is at least one object in this array that cannot be serialized (Parameter 'array')
```

**Comment by tig** [comment]
FYI - your tests may be failing because of work i'm doing on trying to fix #3052. JFYI

**Comment by tig** [comment]
The github action is back to just running on Windows, so you can merge with v2_develop and try again. Sorry about  all this.

**Comment by tznind** [comment]
This is good to go to review, I will look into color picker but that can be seperate PR.

**Comment by tig** [comment]
> This is good to go to review, I will look into color picker but that can be seperate PR.

I have an old pr for that already. Happy to have you contribute to it.

File: Terminal.Gui/Drawing/Gradient.cs [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+    /// <summary>
+    /// Color varies along Y axis but is constant on X axis.
+    /// </summary>
+    Vertical,
+
+    /// <summary>
+    /// Color varies along X axis but is constant on Y axis.
+    /// </summary>
+    Horizontal,
+
```
Comment (tig) [inline Terminal.Gui/Drawing/Gradient.cs:23]: Please use re-sharper's Clean Up Code for all these files.

image

image

File: Terminal.Gui/Drawing/Gradient.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+    Radial,
+
+    /// <summary>
+    /// Color varies by X and Y axis (i.e. a slanted gradient)
+    /// </summary>
+    Diagonal
+}
+
+/// <summary>
+/// Describes
```
Comment (tig) [inline Terminal.Gui/Drawing/Gradient.cs:37]: "Describes" what?

File: Terminal.Gui/Drawing/SolidFill.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+﻿namespace Terminal.Gui.Drawing;
```
Comment (tig) [inline Terminal.Gui/Drawing/SolidFill.cs:1]: For v2 I do want to revisit our namespaces. However, we're not there yet, so can you please make this "Terminal.Gui"?

**Review by tig — CHANGES_REQUESTED** [review]
See my comments. This is lovely.

**Comment by tig** [comment]
It'd be cool if `TextFormatter` had the equivalent of `LineCanvas.FillPair`, no?

File: Terminal.Gui/Drawing/Gradient.cs (hunk 2 above)
Comment (tznind) [reply Terminal.Gui/Drawing/Gradient.cs]: Fixed

File: Terminal.Gui/Drawing/SolidFill.cs (hunk 3 above)
Comment (tznind) [reply Terminal.Gui/Drawing/SolidFill.cs]: Fixed in 1ba84ded9a800206a1724a92980d634522c19bc7 was accident IDae copy paste

File: Terminal.Gui/Drawing/Gradient.cs (hunk 1 above)
Comment (tznind) [reply Terminal.Gui/Drawing/Gradient.cs:23]: I will check when I have access to my windows machine but don't think I have a license for this tool

Comment (tig) [reply Terminal.Gui/Drawing/Gradient.cs:23]: I have licenses for the team. I'll send.

**Comment by tig** [comment]
PR incoming that does this:

AQkKguh 1

**Comment by tig** [comment]
https://github.com/tznind/gui.cs/pull/165/

**Comment by tig** [comment]
Do you think this is ready to merge @tznind. I do.

**Comment by tznind** [comment]
Yup good to go!  

I looked at TextFormatter but looks non trivial.  Can investigate in another pr but I might look to do HSL style true color picker first.

**Comment by tig** [comment]
@tznind please confirm you received email with reshaper license.

**Pull request merged**